### PR TITLE
Use CBOR instead of JSON

### LINKED ISSUE #21: JSON cannot represent Infinity and NaN
>*gets infinity from wpilib.Counter*
>*doesn't show up on dashboard table viewer page*
>*fires up OutlineViewer*
>*infinity*
>*range finder isn't plugged in*

### PR DESCRIPTION
**Pull request by akainth015**
Use CBOR instead of JSON for serializing and de-serializing data because it can handle Infinity or NaN
Note: this will require including `cbor.js` before `networktables.js`. If this is approved, I'll merge the two together, but to make reviewing changes easier they are separate.
Fixes #21

### CODE REVIEW

**Comment by auscompgeek** [comment]
I think, for licensing reasons, cbor.js should remain separate.

**Comment by virtuald** [comment]
What do you mean for licensing reasons?

**Comment by akainth015** [comment]
@auscompgeek The MIT License states that it is acceptable provided the license is included with it.

**Comment by auscompgeek** [comment]
> What do you mean for licensing reasons?

I just think it'll be extremely confusing having portions of the same file under different licenses.

**Comment by akainth015** [comment]
Now that the season is over and the school year is about to end, I think it is a good time to take a look at this pull request, as well as (#36) again, @virtuald.

**Comment by virtuald** [comment]
I guess I didn't give this that much thought. Why CBOR?

**Comment by akainth015** [comment]
Here's the website for CBOR if you want to know more about the specification itself (https://cbor.io/), but the main reasons I had for choosing CBOR is that it is structurally identical to JSON (meaning it wouldn't change our current implementation much), it supports `Infinity` and `NaN` values, and it has support for both JavaScript and Python.

Tangentially it is also faster, but that's because the message is encoded in binary instead of as a string, which has a side-effect of making the message itself non-human-readable. However, I can't imagine any case in which someone would want to do that, because they could just check the pre-serialization object (on the server) or the deserialized object (on the dashboard) instead.

File: pynetworktables2js/nt_serial.py [hunk 1]
```diff
@@ -1,7 +1,8 @@
-try:
-    import ujson as json
-except ImportError:
-    import json
+# try:
+#     import ujson as json
+# except ImportError:
+#     import json
```
Comment (auscompgeek) [inline pynetworktables2js/nt_serial.py]: Please remove this dead code.

File: pynetworktables2js/js/networktables.js [hunk 2]
```diff
@@ -282,7 +282,7 @@ const NetworkTables = new function () {
 			};
 			
 			socket.onmessage = function(msg) {
-				const data = JSON.parse(msg.data);
+				var data = CBOR.decode(msg.data);
```
Comment (auscompgeek) [inline pynetworktables2js/js/networktables.js]: Did you mean to change this back to `var`?

File: pynetworktables2js/js/networktables.js [hunk 3]
```diff
@@ -269,8 +269,8 @@ const NetworkTables = new function () {
 	const address = `${protocol}//${host}/networktables/ws`;
 	
 	function createSocket() {
-	
 		socket = new WebSocket(address);
+    socket.binaryType = "arraybuffer";
```
Comment (auscompgeek) [inline pynetworktables2js/js/networktables.js]: Indentation seems a bit wacky here.

File: pynetworktables2js/js/networktables.js (hunk 2 above)
Comment (akainth015) [reply pynetworktables2js/js/networktables.js]: Nice catch, I'll fix that

**Comment by virtuald** [comment]
Regarding speed: I find it hard to believe that a pure Javascript/python implementation of an encoder/decoder is going to be faster than the C version of a json encoder/decoder provided with python/Javascript. However, I don't think we actually care that much about speed here.

I'm slightly concerned about the cbor python library, it doesn't seem to be maintained anymore and there are a number of reported issues. Additionally, while it has C bindings, they don't publish wheels, so that could cause issues for many of our users.

**Comment by akainth015** [comment]
That's a good point, I didn't consider that the CBOR libraries weren't implemented in C like the JSON ones are. Our team now uses CBOR for our dashboard and we haven't had any noticeable latency changes, so I guess it's true that we shouldn't care about speed.

And I did notice that the python module has not been receiving updates, and I am sure none of us are looking to maintain our own CBOR implementation for this project.

Do you have a suggestion for another serialization format to check out? I can't wrap my head around using a format with a schema properly, so I am looking for formats based off of JSON. BSON seems like a possible alternative (it has both PIP and JavaScript releases), and is actively maintained, although there are no `.whl`s available on PyPI.

**Comment by auscompgeek** [comment]
The cbor PyPI package does include a C implementation, but the open bugs are deeply concerning...

The cbor2 package seems to be well maintained though, and it looks like it'll gain a C extension in the next release.

**Comment by auscompgeek** [comment]
We could also consider protobuf or msgpack, but those seem a bit overkill and would involve a lot more work for this.

**Comment by akainth015** [comment]
I thought about using protocol buffers, but I could not wrap my head around using it for this because the data we transmit is not consistently structured across teams. Can you explain how we'd format data if we used a serialization format with a "schema".

**Comment by auscompgeek** [comment]
`nt_serial` currently defines only 3 messages that are sent between the browser and Python:

- connection status
- incoming update from the NetworkTables server
- incoming update from the browser

Each message would become, well, a message.

**Comment by akainth015** [comment]
Okay, I realize that I was thinking about the data within our

**Comment by auscompgeek** [comment]
I think if we switch this to using the `cbor2` PyPI package then this should be ok. Thoughts?

**Comment by virtuald** [comment]
That seems fine to me.

**Comment by akainth015** [comment]
Hopefully that meets the requirements

**Comment by virtuald** [comment]
@amorygalili can you test this and make sure it doesn't break your thing? If it's fine, then we can merge it.

**Comment by akainth015** [comment]
> 
> 
> @amorygalili can you test this and make sure it doesn't break your thing? If it's fine, then we can merge it.

When you test it, make sure to add a reference to `/networktables/cbor.js`

**Comment by akainth015** [comment]
Are there any updates? @amorygalili @virtuald

File: requirements.txt [hunk 4]
```diff
@@ -1,2 +1,3 @@
+cbor
```
Comment (virtuald) [inline requirements.txt]: cbor2

**Comment by akainth015** [comment]
The only thing missing was `cbor.js`, and I tested with my mock server: https://github.com/akainth015/pynetworktables2js-mock-server (shameless plug, but it will serve as a mocked NetworkTables instance so that you can test your dashboard without access to a robot).

**Comment by virtuald** [comment]
Sorry for the delay, RobotPy has been pretty busy this year. Example works now.

**Pull request merged**